### PR TITLE
New Plant Trait : Bluespace Slips

### PR DESCRIPTION
**Pull request by dootythefrooty**
# Description

Adds a new trait which plants can mutate to have. 
It teleports both the slippee and the produce to a random location within a radius determined by how potent the produce is.
Inert unless the plant also has the slippery trait. 

~~Probably very stinky code considering this is my first time dealing with c#.~~

---
<details><summary><h1>Media</h1></summary>
<p>

https://github.com/user-attachments/assets/cd22756d-ea5e-4a30-8043-c991549c9019

</p>
</details>

---

# Changelog

:cl:
- add: Added Bluespace Slips, a plant trait that teleports you randomly if you slip.

### CODE REVIEW

File: Content.Server/Botany/Systems/TeleportingTraitSystem.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        var randomX = _random.NextFloat(gridBounds.Left, gridBounds.Right);
+        var randomY = _random.NextFloat(gridBounds.Bottom, gridBounds.Top);
+        //probably a better way to do this, but i have no clue at all
+        var otherRandomX = _random.NextFloat(gridBounds.Left, gridBounds.Right);
+        var otherRandomY = _random.NextFloat(gridBounds.Bottom, gridBounds.Top);
+        var producePos = new Vector2(randomX, randomY);
+        var otherPos = new Vector2(otherRandomX, otherRandomY);
+        _xform.SetWorldPosition(uid, producePos);
+        _popup.PopupEntity(Loc.GetString("teleporting-trait-component-slipped"), args.Slipped, args.Slipped, PopupType.SmallCaution);
+        _xform.SetWorldPosition(args.Slipped, otherPos);
```
Comment (VMSolidus) [inline Content.Server/Botany/Systems/TeleportingTraitSystem.cs:47]: _random.NextVector2 lets you create a random vector with a specific magnitude. Set the magnitude to comp.ProduceTeleportRadius. WorldPosition. 
You can also do _xform.SetWorldPosition(uid, Transform(uid).Coordinates + _random.NextVector2(comp.ProduceTeleportRadius, comp.ProduceTeleportRadius))

Don't quote me on if ifs Coordinates or LocalCoordinates being the one you need, you should probably test this. But Transform(uid) just *contains* both LocalCoordinates and WorldCoordinates, you don't need to query it via TransformSystem.

File: Content.Server/Botany/Systems/TeleportingTraitSystem.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    // chance of being deleted and then spawnin the goop
+    private void VanishProbablity(EntityUid uid, TeleportingTraitComponent comp)
+    {
+        if (!_random.Prob(comp.DeletionChance))
+            return;
+        QueueDel(uid);
+        Solution vanishSolution = new();
+        vanishSolution.AddReagent("Slime", comp.Potency / 2);
+        _puddle.TrySpillAt(uid, vanishSolution, out _);
```
Comment (VMSolidus) [inline Content.Server/Botany/Systems/TeleportingTraitSystem.cs:48]:
```suggestion
        if (!_random.Prob(comp.DeletionChance))
            return;
        Solution vanishSolution = new();
        vanishSolution.AddReagent("Slime", comp.Potency / 2);
        _puddle.TrySpillAt(uid, vanishSolution, out _);
        QueueDel(uid);
```
Probably good practice to do the QueueDel after you call a function that requires said entity exists in order to work. :trollface:

File: Content.Server/Botany/Components/TeleportingTraitComponent.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
+namespace Content.Server.Botany
+{
+    [RegisterComponent]
+
+    public sealed partial class TeleportingTraitComponent : Component
+    {
+        ///<summary>
+        /// Teleportation radius of produce.
+        ///
+        [DataField("produceTeleportRadius")]
```
Comment (VMSolidus) [inline Content.Server/Botany/Components/TeleportingTraitComponent.cs:10]:
```suggestion
        [DataField]
```

File: Content.Server/Botany/Components/TeleportingTraitComponent.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        public float ProduceTeleportRadius;
+
+        ///<summary>
+        /// How much to divide the potency.
+        ///
+        [DataField("potencyDivide")]
```
Comment (VMSolidus) [inline Content.Server/Botany/Components/TeleportingTraitComponent.cs:16]:
```suggestion
        [DataField]
```

File: Content.Server/Botany/Components/TeleportingTraitComponent.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+namespace Content.Server.Botany
+{
+    [RegisterComponent]
+
+    public sealed partial class TeleportingTraitComponent : Component
+    {
+        ///<summary>
+        /// Teleportation radius of produce.
+        ///
```
Comment (VMSolidus) [inline Content.Server/Botany/Components/TeleportingTraitComponent.cs:9]:
```suggestion
        /// <summary>
        ///     Teleportation radius of produce.
        /// </summary>
```

File: Content.Server/Botany/Components/TeleportingTraitComponent.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        [DataField("produceTeleportRadius")]
+        public float ProduceTeleportRadius;
+
+        ///<summary>
+        /// How much to divide the potency.
+        ///
```
Comment (VMSolidus) [inline Content.Server/Botany/Components/TeleportingTraitComponent.cs:15]:
```suggestion
        /// <summary>
        ///     How much to divide the potency.
        /// </summary>
```

File: Content.Server/Botany/Components/TeleportingTraitComponent.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        [DataField("potencyDivide")]
+        public float PotencyDivide = 10f;
+
+        ///<summary>
+        ///  Potency of fruit.
+        ///
```
Comment (VMSolidus) [inline Content.Server/Botany/Components/TeleportingTraitComponent.cs:21]:
```suggestion
        /// <summary>
        ///     Potency of fruit.
        /// </summary>
```

File: Content.Server/Botany/Components/TeleportingTraitComponent.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        [DataField("potency")]
```
Comment (VMSolidus) [inline Content.Server/Botany/Components/TeleportingTraitComponent.cs:22]:
```suggestion
        [DataField]
```

File: Content.Server/Botany/Components/TeleportingTraitComponent.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        public float Potency;
+
+        ///<summary>
+        ///  Chance of deletion.
+        ///
```
Comment (VMSolidus) [inline Content.Server/Botany/Components/TeleportingTraitComponent.cs:27]:
```suggestion
        /// <summary>
        ///     Chance of deletion.
        /// </summary>
```

File: Content.Server/Botany/Components/TeleportingTraitComponent.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        [DataField("deletionChance")]
```
Comment (VMSolidus) [inline Content.Server/Botany/Components/TeleportingTraitComponent.cs:28]:
```suggestion
        [DataField]
```

File: Content.Server/Botany/SeedPrototype.cs [hunk 11]
```diff
@@ -205,6 +205,11 @@ public partial class SeedData
     /// </summary>
     [DataField("ligneous")] public bool Ligneous;
 
+    /// <summary>
+    ///     If true, teleports both fruit and player if slippable.
+    /// </summary>
+    [DataField("teleporting")] public bool Teleporting;
```
Comment (VMSolidus) [inline Content.Server/Botany/SeedPrototype.cs:211]:
```suggestion
    [DataField] public bool Teleporting;
```

**Review by VMSolidus — CHANGES_REQUESTED** [review]

File: Content.Server/Botany/Components/TeleportingTraitComponent.cs [hunk 12]
```diff
@@ -0,0 +1,31 @@
+namespace Content.Server.Botany
+{
+    [RegisterComponent]
+
+    public sealed partial class TeleportingTraitComponent : Component
+    {
+        /// <summary>
+        ///     Teleportation radius of produce.
+        ///
```
Comment (VMSolidus) [inline Content.Server/Botany/Components/TeleportingTraitComponent.cs:9]:
```suggestion
        /// </summary>
```

File: Content.Server/Botany/Components/TeleportingTraitComponent.cs [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+        [DataField]
+        public float ProduceTeleportRadius;
+
+        /// <summary>
+        ///     How much to divide the potency.
+        ///
```
Comment (VMSolidus) [inline Content.Server/Botany/Components/TeleportingTraitComponent.cs:15]:
```suggestion
        /// </summary>
```

File: Content.Server/Botany/Components/TeleportingTraitComponent.cs [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+        [DataField]
+        public float PotencyDivide = 10f;
+
+        /// <summary>
+        ///     Potency of fruit.
+        ///
```
Comment (VMSolidus) [inline Content.Server/Botany/Components/TeleportingTraitComponent.cs:21]:
```suggestion
        /// </summary>
```

File: Content.Server/Botany/Components/TeleportingTraitComponent.cs [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+        [DataField]
+        public float Potency;
+
+        /// <summary>
+        ///     Chance of deletion.
+        ///
```
Comment (VMSolidus) [inline Content.Server/Botany/Components/TeleportingTraitComponent.cs:27]:
```suggestion
        /// </summary>
```

**Review by VMSolidus — CHANGES_REQUESTED** [review]

**Review by VMSolidus — APPROVED** [review]

**Review by DangerRevolution — APPROVED** [review]

**Comment by DangerRevolution** [comment]
plant update now where's the zombies?

**Pull request merged**